examples/thompson_sampling.py

```diff
@@ -0,0 +1,302 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Example: Gaussian Process
+=========================
+
+In this example we show how to implement Thompson sampling for Gaussian processes.
+
+"""
+
+import argparse
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+import jax
+import jax.numpy as jnp
+import jax.random as random
+
+import numpyro
+import numpyro.distributions as dist
+from numpyro.infer import SVI, Trace_ELBO
+from numpyro.infer.autoguide import AutoDelta
+
+numpyro.enable_x64()
+
+
+# defining the function. At y=0 to get a 1D cut at the origin
+def ackley_1d(x, y=0):
+    out = (
+        -20 * jnp.exp(-0.2 * jnp.sqrt(0.5 * (x ** 2 + y ** 2)))
+        - jnp.exp(0.5 * (jnp.cos(2 * jnp.pi * x) + jnp.cos(2 * jnp.pi * y)))
+        + jnp.e
+        + 20
+    )
+    return out
+
+
+# matern kernel with nu = 1.5
+def matern32_kernel(X, Z, var=1.0, length=0.5, jitter=1.0e-6):
+    d = jnp.sqrt(0.5) * jnp.sqrt(jnp.power((X[:, None] - Z), 2.0)) / length
+    k = var * (1 + d + (d ** 2) / 3) * jnp.exp(-d)
+    if jitter:
+        k += jitter * jnp.eye(X.shape[0])
+    return k
+
+
+def model(X, Y, kernel=matern32_kernel):
+    # set uninformative log-normal priors on our kernel hyperparameters
+    var = numpyro.sample("var", dist.LogNormal(0.0, 1.0))
+    length = numpyro.sample("length", dist.LogNormal(0.0, 1.0))
+
+    # compute kernel
+    k = kernel(X, X, var, length)
+
+    # sample Y according to the standard gaussian process formula
+    numpyro.sample(
+        "Y",
+        dist.MultivariateNormal(loc=jnp.zeros(X.shape[0]), covariance_matrix=k),
+        obs=Y,
+    )
+
+
+class GP:
+    """ Adapted to numpyro from https://gdmarmerola.github.io/ts-for-bayesian-optim/ """
+
+    def __init__(self, kernel=matern32_kernel):
+        self.kernel = kernel
+        self.kernel_params = None
+
+    def fit(self, X, Y, rng_key, n_step):
+        self.X_train = X
+
+        # store moments of training y (to normalize)
+        self.y_mean = jnp.mean(Y)
+        self.y_std = jnp.std(Y)
+
+        # normalize y
+        Y = (Y - self.y_mean) / self.y_std
+
+        # setup optimizer and SVI
+        optim = numpyro.optim.Adam(step_size=0.005)
+
+        svi = SVI(
+            model,
+            guide=AutoDelta(model),
+            optim=optim,
+            loss=Trace_ELBO(),
+            X=X,
+            Y=Y,
+        )
+
+        params, _ = svi.run(rng_key, n_step)
+
+        # get kernel parameters from guide with proper names
+        self.kernel_params = svi.guide.median(params)
+
+        # store inverted prior covariance
+        self.K_xx_inv = jnp.linalg.inv(self.kernel(X, X, **self.kernel_params))
```

Comment:
given that this code is most of the way towards a complete implementation would it make sense to use `jax.scipy.linalg.cholesky` and related functions instead of `inv`, which tends to be less stable?

Reply:
I'll need to look up the equations again on the weekend, but that's a good idea. Thanks!